QA-1612: TestRunner Enhancement: Consolidate multiple TestRunner JSON outputs into SUMMARY_testRun JSON.

ichengchang opened the pull request:
Consolidate JSON outputs into a single `SUMMARY_testRun.json` file.

Added a GitHub Action to calculate latest Terra service versions that other repos can call.

marikomedlock commented on .github/actions/sync-terra-services-version/action.yml:
```diff
@@ -0,0 +1,104 @@
+        token: ${{ inputs.token }}
+        path: terra-helmfile
+
+    - name: Calculate versions of Terra services
```
Makes sense about including the versions in the dashboard, and using this script as an interim solution. Per your comment below, looking forward to discussing this as separate PR/enhancement.

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/TestScriptResult.java:
```diff
@@ -38,6 +39,24 @@ private TestScriptResultSummary(String testScriptName, String testScriptDescript
+  @SuppressFBWarnings(
+      value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD",
+      justification = "This POJO class is used for easy serialization to JSON using Jackson.")
+  public static class TestScriptUserJourneySnapshots {
```
Ah okay. 2 questions
1. Would it be easier to change the cloud function logic to parse the JSON and insert into BigQuery, rather than modifying the logic here?
2. Can you outline a separate POJO class that has the complete output format you want? (doesn't have to be code, can just be in a comment here). I think that would make it easier to check that we're not duplicating information.

Another related point here. I originally wrote Test Runner library to output a single file. But for runs with lots of metrics readings or hundreds of user journey threads, this single file got unwieldy and started occasionally crashing peoples' text editors. I understand that for automated processing, having things in a single file may make more sense, but I think for developers looking at the files, it would be good to continue writing out separate files. So perhaps we can make this an optional behavior (e.g. continue writing separate files and if a flag is set somewhere, also write the single consolidated file).

ichengchang replied:
The cloud function app that we build (in collaboration with App Services) for ingesting JSON files does not require any application-level parsing logic. The JSON is ingested from the Google bucket as raw bytes and `streaming inserts` of the JSON _key-value_ chunks into `BigQuery` will happen at the correct location of the table schema. All we need is a matching table schema with respect to the `TestRunner` output shape.

Here's the table schema of one row of the `TestRunner` summary:

```
[
  {
    "schema": "[\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"id\",\n    \"type\": \"STRING\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startUserJourneyTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endUserJourneyTime\",\n    \"type\": \"INT64\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"startUserJourneyTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"endUserJourneyTimestamp\",\n    \"type\": \"TIMESTAMP\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"testScriptResultSummaries\",\n    \"type\": \"ARRAY<STRUCT<testScriptDescription STRING, testScriptName STRING, elapsedTimeStatistics STRUCT<min FLOAT64, max FLOAT64, mean FLOAT64, standardDeviation FLOAT64, median FLOAT64, percentile95 FLOAT64, percentile99 FLOAT64, sum FLOAT64>, totalRun INT64, numCompleted INT64, numExceptionsThrown INT64, isFailure BOOL>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"resourceFileName\",\n    \"type\": \"STRING\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"testSuiteName\",\n    \"type\": \"STRING\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"testConfiguration\",\n    \"type\": \"STRUCT<name STRING, description STRING, serverSpecificationFile STRING, billingAccount STRING, isFunctional BOOL, resourceFileName STRING, testUserFiles ARRAY<STRING>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"server\",\n    \"type\": \"STRUCT<name STRING, description STRING, samUri STRING, datarepoUri STRING, samResourceIdForDatarepo STRING, bufferUri STRING, bufferClientServiceAccountFile STRING, bufferClientServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"kubernetes\",\n    \"type\": \"STRUCT<numberOfInitialPods INT64>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"application\",\n    \"type\": \"STRUCT<maxStairwayThreads INT64, maxBulkFileLoad INT64, maxBulkFileLoadArray INT64, loadConcurrentFiles INT64, loadConcurrentIngests INT64, loadDriverWaitSeconds INT64, loadHistoryCopyChunkSize INT64, loadHistoryWaitSeconds INT64>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"testScripts\",\n    \"type\": \"ARRAY<STRUCT<name STRING, numberOfUserJourneyThreadsToRun INT64, userJourneyThreadPoolSize INT64, expectedTimeForEach INT64, expectedTimeForEachUnit STRING, parameters ARRAY<STRING>, expectedTimeForEachUnitObj STRING, description STRING>>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"testUsers\",\n    \"type\": \"ARRAY<STRUCT<name STRING, userEmail STRING, delegatorServiceAccountFile STRING, delegatorServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"disruptiveScript\",\n    \"type\": \"STRUCT<name STRING, parameters ARRAY<STRING>>\"\n  },\n  {\n    \"mode\": \"NULLABLE\",\n    \"name\": \"testConfig\",\n    \"type\": \"STRUCT<name STRING, description STRING, serverSpecificationFile STRING, billingAccount STRING, isFunctional BOOL, resourceFileName STRING, testUserFiles ARRAY<STRING>, server STRUCT<name STRING, description STRING, samUri STRING, datarepoUri STRING, samResourceIdForDatarepo STRING, bufferUri STRING, bufferClientServiceAccountFile STRING, bufferClientServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>, workspaceManagerUri STRING, policyManagerUri STRING, cluster STRUCT<clusterName STRING, clusterShortName STRING, region STRING, zone STRING, project STRING, namespace STRING, containerName STRING, componentLabel STRING, apiComponentLabel STRING>, testRunnerServiceAccountFile STRING, testRunnerServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>, testRunnerK8SServiceAccountFile STRING, testRunnerK8SServiceAccount STRUCT<name STRING, clientKeyDirectoryPath STRING, clientKeyFilename STRING, tokenFilename STRING>, deploymentScript STRUCT<name STRING, parameters ARRAY<STRING>, scriptClass STRING>, skipKubernetes BOOL, skipDeployment BOOL>, kubernetes STRUCT<numberOfInitialPods INT64>, application STRUCT<maxStairwayThreads INT64, maxBulkFileLoad INT64, maxBulkFileLoadArray INT64, loadConcurrentFiles INT64, loadConcurrentIngests INT64, loadDriverWaitSeconds INT64, loadHistoryCopyChunkSize INT64, loadHistoryWaitSeconds INT64>, testScripts ARRAY<STRUCT<name STRING, numberOfUserJourneyThreadsToRun INT64, userJourneyThreadPoolSize INT64, expectedTimeForEach INT64, expectedTimeForEachUnit STRING, parameters ARRAY<STRING>, expectedTimeForEachUnitObj STRING, description STRING>>, testUsers ARRAY<STRUCT<name STRING, userEmail STRING, delegatorServiceAccountFile STRING, delegatorServiceAccount STRUCT<name STRING, jsonKeyDirectoryPath STRING, jsonKeyFilename STRING>>>, disruptiveScript STRUCT<name STRING, parameters ARRAY<STRING>>>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"userJourneySnapshotsCollection\",\n    \"type\": \"ARRAY<STRUCT<testScriptName STRING, testScriptDescription STRING, userJourneyResults ARRAY<STRUCT<userJourneyDescription STRING, threadName STRING, completed BOOL, elapsedTimeNS INT64, exceptionWasThrown BOOL, exceptionStackTrace STRING, exceptionMessage STRING>>>>\"\n  },\n  {\n    \"mode\": \"REQUIRED\",\n    \"name\": \"terraVersions\",\n    \"type\": \"ARRAY<STRUCT<appName STRING, appVersion STRING, chartVersion STRING>>\"\n  }\n]"
  }
]
```

ichengchang replied:
You are right on point that this is meant for automated processing, not something I would expect to be opened in a text editor for inspection. The `BigQuery streaming insert` provides a generous rate limit (up to 10MB per row) and 1GB per second without using `insertId` which is more than sufficient for `TestRunner` across all `MC Terra`. The consolidated summary is <= 10k bytes for `WSM` for each test suite in nightly run.

ichengchang replied:
The cloud function only process `SUMMARY_testRun.json`. Other files like `RAWDATA_userJourneyResults.json` and `RENDERED_testConfiguration.json` still exist in the uploaded `tgz` and can be downloaded. 

Instead of having a switch to get the original shorter `SUMMARY_testRun.json`. How about add a `SUMMARY_testRun_short.json` to the `tgz` as a additional file?

marikomedlock replied:
I like that idea. Slight modification, can you leave the existing filenames in case someone depends on them, and write a new one e.g. `SUMMARY_testRun_concatenated.json`?

marikomedlock replied:
As part of this work, can you pull out the `TestRunSummary` and your new output file JSON format into separate POJO classes? I think it's easier to read if we have all the file formats in their own classes. e.g. for your new format:

```
public static class TestRunSummaryConcatenated {
  public TestRunSummary summary;
  public TestConfiguration config;
  public List<TestScriptResult> testScriptResults;
}
```

ichengchang replied:
> I like that idea. Slight modification, can you leave the existing filenames in case someone depends on them, and write a new one e.g. `SUMMARY_testRun_concatenated.json`?

Yea that's what I intended to comment on but ended up not correcting it when finalizing my comments last Friday. Yes, what I should have said is to keep the original `SUMMARY_testRun.json` and use a new name for the longer version like what you've suggested `SUMMARY_testRun_concatenated.json`.

ichengchang replied:
> As part of this work, can you pull out the `TestRunSummary` and your new output file JSON format into separate POJO classes? I think it's easier to read if we have all the file formats in their own classes. e.g. for your new format:
> 
> ```
> public static class TestRunSummaryConcatenated {
>   public TestRunSummary summary;
>   public TestConfiguration config;
>   public List<TestScriptResult> testScriptResults;
> }
> ```

I did it slightly differently. Instead I created a subclass of `TestRunSummary` 

```
public static class TestRunSummaryConcatenated extends TestRunSummary
```

and use the following views  (@JsonView(SummaryViews.Summary.class) and @JsonView(SummaryViews.ConcatenatedSummary.class)) to generate `SUMMARY_testRun.json` and `SUMMARY_concat_testRun.json` (prefer a shorter name compared to `SUMMARY_testRun_concatenated.json`).

```
public class SummaryViews {
  /** JSON View for TestRunSummary. */
  public static class Summary {}

  /** JSON view for TestRunSummaryConcatenated. */
  public static class ConcatenatedSummary extends Summary {}
}
```

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/TestScriptResult.java:
```diff
@@ -38,6 +39,24 @@ private TestScriptResultSummary(String testScriptName, String testScriptDescript
+      value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD",
+      justification = "This POJO class is used for easy serialization to JSON using Jackson.")
+  public static class TestScriptUserJourneySnapshots {
+    public String testScriptName;
+    public String testScriptDescription;
```
Already there is a `TestScriptResult` class that has 2 properties: `List<UserJourneyResult>` and `TestScriptResultSummary`. The `TestScriptResultSummary` contains both of these properties. So isn't this duplicating the same information in a different object? I'm fine with modifying the output format, but would prefer not to duplicate information.

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/UserJourneyResult.java:
```diff
@@ -19,6 +19,8 @@
   public String exceptionStackTrace;
   public String exceptionMessage;
 
+  public UserJourneyResult() {}
```
Just curious, why was this needed?

ichengchang replied:
I think that's because `jackson` complained about the need of the no-arg constructor when I added `public TestScriptUserJourneySnapshots userJourneySnapshots` to `TestScriptResult` class.

marikomedlock reviewed (commented):
Two main questions about: moving version-related changes to a separate PR, de-duplicating information in the output format. Maybe would be good to discuss in person over GVC?

ichengchang commented on src/main/java/bio/terra/testrunner/runner/SummaryViews.java:
```diff
@@ -0,0 +1,9 @@
+package bio.terra.testrunner.runner;
+
+public class SummaryViews {
```
You are correct that the two approaches do exactly the same thing. The first one as you suggested wraps `TestRunSummary` while the second one as currently implemented extends `TestRunSummary` by the same set of properties. 

The point is that approach 1 has a different _schema_ than approach 2 whereas the `testRunSummary` becomes an entity directly under the root and all the `TestRunSummary` properties appears at the next level down. All the `TestRunSummary` properties are at the root level in approach 2.

If we use approach 1, we would have to make schema change to `BigQuery` in addition to moving existing `WSM` test data to fit the new schema. In approach 2, we just need to extend the `BigQuery` schema with the concatenated properties so no need to touch existing data from `TestRunSummary`.

Here's where the Json View comes in when we use the 2nd approach. Casting a Java subclass to its superclass (from `TestRunSummaryConcatenated` into `TestRunSummary`) makes no difference when comes to Json serialization. When done naively, the serialization still projects the full `TestRunSummaryConcatenated` from memory. The Json View correctly _filters_ properties based on the different views to get the desired output, in this case the full `TestRunSummaryConcatenated` and the shorter version `TestRunSummary`.

So really the benefits are

- Minimal change to existing `BigQuery` schema.
- Better organization and flexibility in managing the reporting of `TestRunner` outputs ( can potentially use different views for different purpose).

**Approach 1**
```
  public static class TestRunSummaryConcatenated {
    private TestRunSummary testRunSummary;
    private TestConfiguration testConfig;
    private List<TestScriptResult.TestScriptUserJourneySnapshots> userJourneySnapshotsCollection;
    private List<TerraVersion> terraVersions;
```

**Approach 2**
```
  public static class TestRunSummaryConcatenated extends TestRunSummary {
    @JsonView(SummaryViews.ConcatenatedSummary.class)
    private TestConfiguration testConfig;

    @JsonView(SummaryViews.ConcatenatedSummary.class)
    private List<TestScriptResult.TestScriptUserJourneySnapshots> userJourneySnapshotsCollection;

    @JsonView(SummaryViews.ConcatenatedSummary.class)
    private List<TerraVersion> terraVersions;
```

ichengchang commented on src/main/java/bio/terra/testrunner/runner/config/KubernetesServiceAccountSpecification.java:
```diff
@@ -12,8 +13,9 @@
-  public File clientKeyFile;
-  public File tokenFile;
+
+  @JsonIgnore public File clientKeyFile;
+  @JsonIgnore public File tokenFile;
```
Yea, these review the location of sensitive information and we don't need these in the reports.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -44,13 +47,24 @@
   private boolean exceptionThrownInCleanup = false;
 
+  /** A base class with test runner summary statistics. Suitable for human inspection. */
   public static class TestRunSummary {
```
Refactored `TestRunSummary` and `TestRunSummaryConcatenated` into separate files.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -100,15 +122,69 @@ private static String millisecondsToTimestampString(long milliseconds) {
-
-    this.summary = new TestRunSummary(UUID.randomUUID().toString());
+    this.summaryConcatenated = new TestRunSummaryConcatenated(UUID.randomUUID().toString());
+    summaryConcatenated.setTerraVersions(TerraVersion.loadEnvVars());
```
I think the bottomline is we need to be able to tell which Terra versions are involved in a test run for triage and audit purpose. The app/chart approach is not perfect and `DevOps` is working on an API solution. I don't know the answer for Verily deployments. I think we should discuss with `DevOps` what their API solution entail with respect to version discovery.

The only reason I included `TerraVersion` here is that QA is trying to accomplish rendering of `WSM` test results in the new dashboard with user journey test summaries along with version information, as part of our Q4 OKRs. Knowing that `TerraVersion`  is far from perfect.

ichengchang commented on src/main/java/bio/terra/testrunner/common/TerraVersion.java:
```diff
@@ -0,0 +1,102 @@
+@SuppressFBWarnings(
+    value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD",
+    justification = "This POJO class is used for easy serialization to JSON using Jackson.")
+public class TerraVersion {
```
See my comments below regarding our need for accomplishing the QA Q4 OKRs and the uncertainty on when the version API `DevOps` is working on will be ready. Hence my preference is to include version-related stuff in this PR. If what you mean is to just have a separate ticket and PR for version-related stuff right after this PR that's fine with me.

ichengchang commented:
@marikomedlock  I copied most of your code from [PF-1131](https://github.com/DataBiosphere/terra-test-runner/pull/23) into this PR. The only updates I made is the following

`public class TestRunFullOutput extends TestRunSummary`

based on our discussion last week. This ensures that the `TestRunSummary id` property corresponds to the unique id field in `BigQuery`.

I am thinking that we should first merge your [PF-1130](https://github.com/DataBiosphere/terra-test-runner/pull/22) into `main` after you review and approve my changes. Afterwards we should focus on this PR and replace `TerraVersion` with the `VersionScript` from PF-1130.

Let's have another meeting to go over this.

marikomedlock commented:
> I am thinking that we should first merge your [PF-1130](https://github.com/DataBiosphere/terra-test-runner/pull/22) into `main` after you review and approve my changes. Afterwards we should focus on this PR and replace `TerraVersion` with the `VersionScript` from PF-1130.
> 
> Let's have another meeting to go over this.

@ichengchang 
That makes sense to me. Look forward to talking about.

ichengchang commented:
@marikomedlock Here's the finalized version of the PR. Refactored the code to use `VersionScript`. I have updated the `determine` method to return a list of `VersionScriptResult` objects. Also added a new property `appName` to `VersionScriptResult` to keep track of the MCTerra service.

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -35,79 +33,25 @@
   private VersionScriptResult versionScriptResult;
   private List<TestScriptResult> testScriptResults;
   protected TestRunSummary summary;
+  protected TestRunFullOutput runFullOutput;
```
This is confusing to me because now we're tracking output in multiple places (e.g. `testScriptResults` and `runFullOutput.testScriptResults`). Instead, could we just create a `TestRunFullOutput` object right before we write out the file? i.e. in `writeOutResults` method, call the constructor and pass in all the relevant `TestRunner` properties: `new TestRunFullOutput(config, summary, testScriptResults, versionScriptResult)`.

Then in `TestRunFullOutput` POJO class, you can unpack those or reformat them however we want for the full output format. In that case, I think you would also not need to have `TestRunFullOutput` extend `TestRunSummary`. They can be completely separate serialization formats.:
```
public class TestRunFullOutput {
  public TestConfiguration testConfiguration;
  public List<TestScriptResult> testScriptResults;
  public VersionScriptResult versionScriptResult;
    public String id;
    public long startTime;
    ...(and other properties)...

  public TestRunFullOutput(
      TestConfiguration testConfiguration,
      TestRunSummary testRunSummary,
      List<TestScriptResult> testScriptResults,
      VersionScriptResult versionScriptResult) {
    this.testConfiguration = testConfiguration;
    this.testScriptResults = testScriptResults;
    this.versionScriptResult = versionScriptResult;

    // unpack the summary object
    this.id = summary.id;
    this.startTime = summary.startTime;
    this.testScriptResultSummaries = summary.testScriptResultSummaries;
    ...(and other properties)...
  }
}
```

Or another way of looking at it -- let's have `TestRunFullOutput` be strictly a serialization format. No need to use it internally in `TestRunner`, just when we write output files.

ichengchang replied:
I forgo the use of `@JsonView` since it appears to create quite a bit of confusion in this PR. I removed the @JsonView annotation and instead copy the `TestRunSummary` properties directly inside `TestRunFullOutput`.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/VersionScript.java:
```diff
@@ -17,7 +17,7 @@ public VersionScript() {}
 
   /** The version script determineVersion method looks up the version. */
-  public VersionScriptResult determineVersion(ServerSpecification server) throws Exception {
+  public List<VersionScriptResult> determineVersion(ServerSpecification server) throws Exception {
```
Reverted `VersionScriptResult` to singular but added a list of `HelmVersion` and Helm-specific version builder. Other version source can have their own specific version builder and overloaded `add` method to build version list.

ichengchang commented on .github/actions/sync-terra-services-version/action.yml:
```diff
@@ -0,0 +1,110 @@
+        # Output Terra versions to $GITHUB_ENV with env names in the following format:
+        #   <appName>_appVersion, <appName>_chartVersion
+        # e.g. workspacemanager_appVersion, workspacemanager_chartVersion
+        calculate_terra_versions () {
```
The `action.yml` has been removed as `VersionScript` takes over.

ichengchang commented on .github/actions/sync-terra-services-version/action.yml:
```diff
@@ -0,0 +1,110 @@
+          done < <(yq e '.releases | keys' terra-helmfile/$VER_FILE_PATH -o=j | jq -r -c '.[]')
+        }
+
+        # Determine the base version file to use and load versions into envars.
```
The `action.yml` has been removed as `VersionScript` takes over.

ichengchang commented on .github/actions/sync-terra-services-version/action.yml:
```diff
@@ -0,0 +1,110 @@
+        fi
+
+        # Refer tp https://github.com/mikefarah/yq#install
+        # Install yq if it is not already installed.
```
The `action.yml` has been removed as `VersionScript` takes over.

ichengchang commented:
@marikomedlock In the latest changes. I have removed `action.yml` and refactored `VersionScriptResult` with a list of `HelmVersion` objects. The `determine` method in version spec returns the original singular `VersionScriptResult`.

ichengchang commented:
The results of a test run can be downloaded from [https://console.cloud.google.com/storage/browser/_details/terra-kernel-k8s-testrunner-results/83967fd9-5a32-485b-9928-0c56903b687b.tar.gz;tab=live_object?cloudshell=true&project=terra-kernel-k8s](https://console.cloud.google.com/storage/browser/_details/terra-kernel-k8s-testrunner-results/83967fd9-5a32-485b-9928-0c56903b687b.tar.gz;tab=live_object?cloudshell=true&project=terra-kernel-k8s)

ichengchang commented:
@marikomedlock  I made a few more updates. I am thinking since we are at it why not include support for git hash as well. We can do a quick sync up today and resolve any outstanding questions.

Added support for Git hash. Here's the new files
`scripts/versionscripts/model/GitRelease`
`scripts/versionscripts/ReadFromGitCommitLog`
`scripts/versionscripts/ReadFromTerraHelmFileRepoAndGitCommitLog`

Updated `ServerSpecification` to accept multiple `VersionScriptSpecification` objects.
Updated `TestRunFullOutput` to return multiple `VersionScriptResult` objects.

You can see a test run [here](https://github.com/DataBiosphere/terra-workspace-manager/runs/4189951865?check_suite_focus=true).

Here's how the server spec looks like

```
{
  "name": "workspace-wsmtest",
  "description": "Workspace Manager Test Environment",

  "samUri": "https://sam.dsde-dev.broadinstitute.org",
  "datarepoUri": "https://jade.datarepo-dev.broadinstitute.org/",
  "workspaceManagerUri": "https://workspace.wsmtest.integ.envs.broadinstitute.org/",

  "cluster": {
    "clusterName": "gke_terra-kernel-k8s_us-central1-a_terra-integration",
    "clusterShortName": "terra-integration",
    "region": "us-central1",
    "zone": "us-central1-a",
    "project": "terra-kernel-k8s",
    "namespace": "terra-wsmtest",
    "containerName": "workspacemanager-deployment",
    "apiComponentLabel":  "workspacemanager"
  },
  "deploymentScript": {},
  "testRunnerServiceAccountFile": "testrunner-sa.json",
  "testRunnerK8SServiceAccountFile": "testrunner-k8s-wsmtest-sa.json",

  "skipDeployment": true,
  "skipKubernetes": false,

  "versionScripts": [
    {
      "name": "ReadFromTerraHelmFileRepoAndGitCommitLog",
      "description": "Version from https://github.com/broadinstitute/terra-helmfile and Git Commit Log",
      "parameters": ["terra-helmfile/versions/app/dev.yaml", "terra-helmfile/environments/personal/wsmtest.yaml", "../.git"]
    }
  ]
}
```

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -394,6 +338,39 @@ private void executeTestConfigurationNoGuaranteedCleanup() throws Exception {
+  private VersionScriptResult determineVersion(
+      VersionScript versionScript, ServerSpecification server) {
+    try {
+      return versionScript.determineVersion(server);
```
Similar to the previous comment, I think you don't need to wrap this in a `RuntimeException`. Can you just call them both directly in the method above? e.g.:
```
      versionScriptResults =
          versionScripts.stream()
              .map(
                  versionScript -> {
                    logger.info(
                        "Version: Calling {}.determineVersion()",
                        versionScript.getClass().getName());
                    versionScript.setParameters(versionScriptSpecification.parameters);
                    return versionScript.determineVersion(versionScript, config.server);
                  })
              .collect(Collectors.toList());
```

ichengchang replied:
Simplified code with a for-loop.

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -394,6 +338,39 @@ private void executeTestConfigurationNoGuaranteedCleanup() throws Exception {
+          "Error calling constructor of VersionScript class: " + versionScriptSpecification.name,
+          e);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
```
I don't think you need to wrap in a `RuntimeException` here. Can you just call the `setParameters` inline above?

ichengchang replied:
Simplified code with a for-loop.

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/VersionScriptResult.java:
```diff
@@ -1,49 +1,129 @@
+  // TODO: add more overloaded add methods for other version sources
+
+  // Builder for helm versions
+  public static class HelmVersion {
```
nit: In keeping with your general cleanup here and separating POJO classes into their own files, could you also move `HelmVersion` and `GitVersion` into separate classes? Perhaps they can go in a new `runner.version` package. I think that will probably get refactored when we make the `VersionScriptResult` class overridable by different `VersionScript`s, but this is good for now.

ichengchang replied:
Refactored `HelmVersion, GitVersion` into separate classes.

marikomedlock commented on src/main/java/scripts/versionscripts/ReadFromGitCommitLog.java:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  /**
+   * This method determines the version by reading the helm app/chart versions in the terra-helmfile
+   * GitHub repository.
```
nit: copy/paste mismatched comment

ichengchang replied:
Fixed comments.

marikomedlock commented on src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java:
```diff
@@ -66,38 +64,11 @@ public VersionScriptResult determineVersion(ServerSpecification server) throws E
-    return to;
+    return Arrays.asList(
+        new VersionScriptResult.HelmVersion.Builder()
+            .appName("workspacemanager")
```
Since this is just for WSM, perhaps you can take the app name as a parameter to this script? That way we could reuse this for other apps also, since I expect the helm merge logic would be identical.

ichengchang replied:
Replaced hardcoded value by `appName` property.

marikomedlock commented on src/main/java/scripts/versionscripts/ReadFromTerraHelmFileRepoAndGitCommitLog.java:
```diff
@@ -0,0 +1,49 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReadFromTerraHelmFileRepoAndGitCommitLog extends VersionScript {
```
Instead of combining the two version lookups in a single script, can you take advantage of your change to allow multiple version scripts per server? i.e. use both `ReadFromGitCommitLog` and `ReadFromTerraHelmfileRepo`

ichengchang replied:
Multiple version script already supported in latest `ServerSpecification`.

marikomedlock replied:
Do you still need this class?

ichengchang replied:
Removed `ReadFromTerraHelmFileRepoAndGitCommitLog` class.

marikomedlock reviewed (commented):
This looks great, thanks for the changes. The Git hash lookup is awesome! I think that will be helpful for the Verily deployments too, though I know that's not your priority here.

Mostly nits. Main comment is whether you can use 2 version scripts instead of combining them into one.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/TestRunFullOutput.java:
```diff
@@ -0,0 +1,61 @@
+ * <p>This extended version is suitable for Cloud Function ingestion of test results as single rows
+ * and for streaming inserts to BigQuery in bulk.
+ */
+/**
```
Merged both comments into one.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/TestRunFullOutput.java:
```diff
@@ -0,0 +1,61 @@
+  public List<TestScriptResult> testScriptResults;
+  public List<VersionScriptResult> versionScriptResults;
+  public String id;
+  public long startTime = -1;
+  public long startUserJourneyTime = -1;
+  public long endUserJourneyTime = -1;
+  public long endTime = -1;
```
Let's keep the `long` values for now. Once the `dashboard` is functioning, I will create a ticket for general clean up task.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -394,6 +338,39 @@ private void executeTestConfigurationNoGuaranteedCleanup() throws Exception {
+      versionScript.setParameters(versionScriptSpecification.parameters);
+
+      // call the determineVersion method to get the version
+      logger.info("Version: Calling {}.determineVersion()", versionScript.getClass().getName());
```
Updated comments in `instantiateVersionScript` method.

ichengchang commented:
@marikomedlock Addressed all your comments. 

Here's a test run

[https://github.com/DataBiosphere/terra-workspace-manager/runs/4218188538?check_suite_focus=true](https://github.com/DataBiosphere/terra-workspace-manager/runs/4218188538?check_suite_focus=true)

Here's the version results

```
[ {
  "helmVersions" : [ {
    "appName" : "workspacemanager",
    "helmAppVersion" : "0.254.125",
    "helmChartVersion" : "0.48.0"
  } ],
  "gitVersions" : null
}, {
  "helmVersions" : null,
  "gitVersions" : [ {
    "remoteOriginUrl" : "https://github.com/DataBiosphere/terra-workspace-manager",
    "branch" : "QA-1612",
    "refHeadCommit" : "a22045da9d1afd29e23860efa05c094fcb165d41",
    "shortRefHeadCommit" : "a22045d"
  } ]
} ]
```

ichengchang commented:
@marikomedlock  I think we can ignore `public List<VersionScriptSpecification> versionScripts;` in the `ServerSpecification` json output since we only care about the `VersionScriptResult`. What do you think?

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/TestScriptResult.java:
```diff
@@ -2,41 +2,13 @@
-      this.testScriptDescription = testScriptDescription;
-    }
-  }
+  @JsonIgnore public TestScriptResultSummary summary;
```
I think you don't want `@JsonIgnore` here. RE your comment in the main PR, `SUMMARY_testRun.json` contains the `TestRunSummary`, which is different than the `TestScriptResultSummary` here.

ichengchang replied:
Replaced `testResultScripts` in `TestRunFullOutput` with List<Map<String, List<UserJourneyResult>> object. Effect is the same as using `@JsonIgnore`. So removed `@JsonIgnore` annotation.

marikomedlock replied:
Sorry, I'm not following why you want to omit `TestScriptResultSummary` objects from the full JSON output. Don't you want information from those objects in the dashboard? Those objects hold all the statistics calculated about each set of test scripts (e.g. average time per user journey, 99 percentile time per user journey).

ichengchang replied:
The `TestRunFullOutput` already contains a list of `TestScriptResultSummary` results

`public List<TestScriptResultSummary> testScriptResultSummaries;`

and the summaries are populated by passing the `TestRunSummary` in the constructor method

```
  public TestRunFullOutput(
      TestConfiguration testConfiguration,
      List<Map<String, List<UserJourneyResult>>> testScriptResults,
      TestRunSummary testRunSummary,
      List<VersionScriptResult> versionScriptResults) {
...
   this.testScriptResultSummaries = testRunSummary.testScriptResultSummaries;
```

So I don't want to duplicate the summaries in testScriptResults again.

Let's have a zoom call if this isn't clear.

marikomedlock replied:
Ah I see. Would it be easier to just remove the `TestRunFullOutput.testScriptResultSummaries` property instead?

marikomedlock replied:
Gotcha. Agreed we should do whatever makes sense for the dashboard in the `TestRunFullOutput` class, since that's its main purpose.

Not sure I'm following the `List<Map<String, List<UserJourneyResult>>>` though -- aren't you just creating a list of maps with a single identical key `userJourneyResults`? Isn't that just a list of `List`s -- why do you need the `Map`?

It feels like we're missing a mapping here though. If you're looking at a `TestScriptResultSummary`, there's no way to link that back to the `UserJourneyResult`s for that same `TestScript`. I think we didn't have this problem before because the `TestScriptResultSummary` and the `List<UserJourneyResult>` are in the same `TestScriptResult` object. Now that you'll separate them, do we need a test script id or something to be able to link them? What do you think?

ichengchang replied:
Yes we can link back the `UserJourneyResult`s to the corresponding `TestScriptResultSummary`. I added the following comments to the `TestRunFullOutput`. Using the position argument would be sufficient.

```
  // The testScriptResults and testScriptResultSummaries must have identical size.
  // For each element in testScriptResults, we can derive the testScriptName and
  // testScriptDescription properties from the corresponding positional element in
  // testScriptResultSummaries.
  public List<Map<String, List<UserJourneyResult>>> testScriptResults;
  public List<TestScriptResultSummary> testScriptResultSummaries;
```

ichengchang replied:
Another more flexible is to have an explicit `unnestUserJourneyResults` method that can shape the object we desire for automated reporting. I just implemented this method and checked in the code,

ichengchang replied:
Regarding your other comments. Quote

```
Not sure I'm following the List<Map<String, List<UserJourneyResult>>> though -- aren't you just creating a list of maps with a single identical key userJourneyResults? Isn't that just a list of Lists -- why do you need the Map?
```

List of **maps** is easier to deal with than list of **lists**. For list of maps, I can unnest the top-level list first and access the list of user journeys by the key name. For list of lists, I ran into trouble using indices to access the result. ORM (SQLAlchemy) works better but I haven't tried out yet.

marikomedlock replied:
Makes sense. Perhaps include that as a comment? I think all these troubles you're running into creating a BQ easily-queryable schema are good information for the next person who has to modify this.

marikomedlock commented on src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java:
```diff
@@ -29,75 +28,49 @@ public ReadFromTerraHelmfileRepo() {}
-    String wsmHelmChartVersion =
-        helmRelease.getReleases().get("workspacemanager").getChartVersion().orElse("");
+    // Pull appName versions from terra-helmfile, add more service versions if needed.
+    String helmAppVersion = helmRelease.getReleases().get(appName).getAppVersion().orElse("");
```
nit: perhaps log if the appName/appVersion is not found?

ichengchang replied:
Added log statements for empty app or chart version.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -30,83 +29,28 @@
+  private List<VersionScriptResult> versionScriptResults;
   private List<TestScriptResult> testScriptResults;
   protected TestRunSummary summary;
+  protected TestRunFullOutput runFullOutput;
```
Removed unused property.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/version/GitVersion.java:
```diff
@@ -0,0 +1,52 @@
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+// Builder for git versions
```
Moved comments to Builder class.

ichengchang commented on src/main/java/bio/terra/testrunner/runner/version/HelmVersion.java:
```diff
@@ -0,0 +1,44 @@
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+// Builder for helm versions
```
Moved comments.

ichengchang commented:
@marikomedlock All comments have been addressed. Thanks for reviewing!

marikomedlock commented on src/main/java/bio/terra/testrunner/runner/TestRunner.java:
```diff
@@ -513,6 +511,27 @@ protected void writeOutResults(String outputParentDirName) throws IOException {
   }
 
+  // A mapper that unnest List<TestScriptResult> from TestScriptResult
+  private List<Map<String, Object>> unnestUserJourneyResults(
```
This is much easier to follow, thanks! Can you move this to the `TestRunFullOutput` class and call it e.g. from the constructor instead of adding to `TestRunner`?

ichengchang replied:
Refactored `unnestUserJourneyResults` into `TestRunFullOutput`. Below is the output of the `TestRunFullOutput.testScriptResults` upon a test run.

```
  "testScriptResults" : [ {
    "userJourneyResults" : [ {
      "userJourneyDescription" : "ServiceStatus: 150",
      "threadName" : "pool-2-thread-1",
      "completed" : true,
      "elapsedTimeNS" : 150608732872,
      "exceptionWasThrown" : false,
      "exceptionStackTrace" : null,
      "exceptionMessage" : null
    }, {
      "userJourneyDescription" : "ServiceStatus: 150",
      "threadName" : "pool-2-thread-2",
      "completed" : true,
      "elapsedTimeNS" : 150608761870,
      "exceptionWasThrown" : false,
      "exceptionStackTrace" : null,
      "exceptionMessage" : null
    } ],
    "testScriptName" : "ServiceStatus",
    "testScriptDescription" : "ServiceStatus: 150"
  } ],

```

marikomedlock reviewed (approved):
LGTM. Thanks for working through this with me!

One future task (either in this PR or perhaps add it to the future work around `VersionScriptResult`): Add documentation about `VersionScript`s to the README.

ichengchang commented:
> LGTM. Thanks for working through this with me!
> 
> One future task (either in this PR or perhaps add it to the future work around `VersionScriptResult`): Add documentation about `VersionScript`s to the README.

Yup that's on my TODO list to create the future README task.
